# Clase 2: Estructuras condicionales y ciclos
Notebook desarrollado por los docentes Andres Castro y Leonardo Duque. Modificado por Cristian Guarnizo.
### 1. Operadores de Asignación

### 2. Operadores Relacionales y Operadores Lógicos

### 3. Estructuras condicionales:

- <font color="#130eb3">__if__</font>
- <font color="#130eb3">__if__ __else__</font>
- <font color="#130eb3">__elif__</font>
- <font color="#130eb3">__if__</font> anidados

### 4. Estructura match (switch-case):

- <font color="#130eb3">__match__</font>





# 1. Operadores de Asignación

Son los símbolos que se utilizan para darle un valor a una variable o para actualizar el valor que ya tiene.

In [ ]:
a = 10

a += 7     # a = a + 7     Suma
a -= 12    # a = a - 12    Resta
a *= 6     # a = a * 6     Multiplicación
a /= 10    # a = a / 10    División
a **= 5    # a = a ** 5    Potencia
a //= 2    # a = a//2      División de piso (dividir los números y redondear el resultado hacia abajo)
a %= 2     # a = a % 2     Modulo o residuo de la division. Es el "sobrante" de una división entera.

print(f"valor de a = {a}")

In [ ]:
a = 10
a

10

In [ ]:
a += 7     # a = a + 7
a

17

In [ ]:
a -= 12    # a = a - 12
a

5

In [ ]:
a *= 6     # a = a * 6
a

30

In [ ]:
a /= 10    # a = a / 10
a

3.0

In [ ]:
a **= 5    # a = a ** 5
a

243.0

In [ ]:
a //= 2    # a = a//2  División de piso (dividir los números y redondear el resultado hacia abajo)
a

121.0

In [ ]:
a %= 2    # a = a % 2 Modulo o residuo de la division. Es el "sobrante" de una división entera.
a

1.0


# 2. Operadores Relacionales en Python


- Se usan para establecer relación entre valores (variables)
- Producen un resultado booleano (True o False):

Una expresión Booleana se puede definir como toda expresión que retorna un valor falso o verdadero:
     
$ expresion \rightarrow [True, False] $

- Todos tienen el mismo nivel de precedencia (prioridad), pero son de menor prioridad que los operadores aritméticos

| Operador | Operación |
| --- | --- |
| > | Mayor que |
| < | Menor que |
| >= | Mayor o igual que |
| <= | Menor o igual que |
| != | Diferente  |
| == | Igual |

In [ ]:
# Celda para ejemplos de operadores relacionales

a = 10
b = 3

variable = a >= b

print(variable)
print(type(variable))

True
<class 'bool'>


In [ ]:
# Los operadores se pueden encadenar
print(0 < a < 10)

True


In [ ]:
x = 0 < a < 10
print(x)

True


# 2. Operadores Lógicos en Python


- Se usan para establecer relación entre valores (variables)
- Producen un resultado booleano (True o False):

| Operador | Operación | Resultado |
| --- | --- |--- |
| and | Y (conjunción) | True cuando ambas condicones son verdaderas |
| or | O (disyunción) | True cuando al menos una condición es verdadera |
| not | No (negación) | Invierte el valor |


In [ ]:
a = 3

print( a > 10 and a < 20)

print( a > 1 and a < 20)

False
True


In [ ]:
print( a > 10 or a < -10)

print( a > 10 or a < 10 )

False


True

In [59]:
print(not( a > 10 or a < -10))

not( a > 10 or a < 10)

True


False

# 3. Estructura condicionales

En Python contamos con tres elementos condicionales de la siguiente forma:

1. **if**(_expresion booleana_)**:** --> Si se cumple la expresión booleana.
2. **else:** --> en caso contrario.
3. **elif**(_expresion booleana_)**:** --> Si no se cumple la expresión anterior y se cumple esta. Viene a reemplazar el `if` anidado de un nivel.


# IF

Si la condición que pones al lado es verdadera (True), el código dentro de él se ejecuta. Si no, se lo salta.

<br>

<font color="#130eb3" size=5>__if__</font> <font size=5>(condición)</font> <font color="#130eb3" size=5>__:__</font>

    instrucción 1
    
    instrucción 2
    
    ...
    
    instrucción n


In [64]:
#Ejemplo if

edad = float(input("Ingrese su edad en años: "))

if(edad >= 18):
    print("Eres mayor de edad")
    print("instruccion 2")

print("Esto es fuera del if")

Ingrese su edad en años: 10
Esto es fuera del if



# IF - ELSE

Si no se cumple la condición del IF (o cualquier otra condición anterior), se ejecuta automáticamente. Es opcional, pero útil para cerrar el ciclo.

<br>

<font color="#130eb3" size=5>__if__</font> <font size=5>(condición)</font> <font color="#130eb3" size=5>__:__</font>

    instrucción 1
    
    instrucción 2
    
    ...
    
    instrucción n
    
<font color="#130eb3" size=5>__else:__</font>

    instrucción sino 1
    
    instrucción sino 2
    
    ...
    
    instrucción sino n

No lleva ninguna condición porque se ejecuta automáticamente si ninguna de las condiciones anteriores fue verdadera. Es opcional, pero útil para cerrar el ciclo.

In [101]:

#Ejemplo if - else

edad = float(input("Ingrese su edad en años: "))

if edad >= 18:
    print("Eres mayor de edad")
else:
    print("No eres mayor de edad")

print("Esto es fuera del if-else")

Ingrese su edad en años: 50
Eres mayor de edad
Esto es fuera del if-else


# ELIF (ELSE - IF )

Se usa cuando quieres revisar una segunda, tercera, cuarta... condición(es), pero solo si las anteriores fallaron. Puedes tener tantos elif como necesites.

<br>

<font color="#130eb3" size=5>__if__</font> <font size=5>(condición)</font> <font color="#130eb3" size=5>__:__</font>

    instrucción 1
    
    instrucción 2
    
    ...
    
    instrucción n
    
<font color="#130eb3" size=5>__elif__</font> <font size=5>(condición2)</font> <font color="#130eb3" size=5>__:__</font>

    instrucción condición2 1
    
    instrucción condición2 2
    
    ...
    
    instrucción condición2 n
    
<font color="#130eb3" size=5>__else:__</font>

    instrucción sino 1
    
    instrucción sino 2
    
    ...
    
    instrucción sino n

In [71]:
#Ejemplo elif - else

edad = float(input("Ingrese su edad en años: "))

if (edad < 18):
    print("No eres mayor de edad")
elif (edad >= 60):
    print("Eres un adulto mayor")
else:
    print("Eres mayor de edad")

print("Esto es fuera del if-else")

Ingrese su edad en años: 65
Eres un adulto mayor
Esto es fuera del if-else


In [88]:
# Ejemplo if-elif-else
# Solicitar coordenadas de un punto y decir en qué cuadrante del plano cartesiano se encuentra
# (x, y)

x = float(input("ingrese el valor de x: "))
y = float(input("ingrese el valor de y: "))

if (x>0):
    if (y<0):
        print("El punto está en el cuadrante 4")
    else:
        print("El punto está en el cuadrante 1")
else:
    if (y<0):
        print("El punto está en el cuadrante 3")
    else:
        print("El punto está en el cuadrante 2")

ingrese el valor de x: 5
ingrese el valor de y: 2
El punto está en el cuadrante 1


In [ ]:
# Ejemplo if-elif-else
# Solicitar coordenadas de un punto y decir en qué cuadrante del plano cartesiano se encuentra
# (x, y)

x = float(input("ingrese el valor de x: "))
y = float(input("ingrese el valor de y: "))

if (x > 0 and y > 0):
    print("El punto está en el cuadrante 1")
elif (x < 0 and y > 0):
    print("El punto está en el cuadrante 2")
elif (x < 0 and y < 0):
    print("El punto está en el cuadrante 3")
elif (x == 0 and y == 0):
    print("El punto está en el origen")
elif (x == 0):
    print("El punto está sobre el eje horizontal")
elif (y == 0):
    print("El punto está sobre el eje vertical")
else:
    print("El punto está en el cuadrante 4")

ingrese el valor de x: 1
ingrese el valor de y: 0
El punto está sobre el eje vertical


## **Ejercicio 1:** Ingresar un numero por consola, decir si es positivo, negativo o cero

In [92]:
# ingresar un numero por consola, decir si es positivo, negativo o cero
num = float(input("ingrese un número: "))

#  Hacer codigo aqui


ingrese un número: 0
El nùmero es cero


# 1.3. Inline IF

Se utiliza para asignar el resultado de una selección condicional a una variable.

In [102]:
num = int(input("ingrese un número entero: "))

#Inline if
print('numero par') if num %2 == 0 else print("numero impar")

# La estructura es la siguiente:
# if num%2 == 0:
#     print('numero par')
# else:
#     print('numero impar')

ingrese un número entero: 6
numero par


In [106]:
x = 20

'''
Si x es menor que 10, asigne x al cuadrado a la variable msg,
en caso contrario, asigne el valor 1 sobre x a la variable msg
'''

# if (x < 10):
#     msg = x*x #x**2
# else:
#     msg = 1/x

# print(msg)

msg = x*x if(x < 10) else 1/x

print(msg)

0.05


## **Ejercicio 2:** Solicitar un valor e imprimir el valor absoluto de él (Usar inline IF)

In [ ]:
# Solicitar un valor e imprimir el valor absoluto de él sin usar abs
num = float(input("Ingrese un número: "))

#Hacer codigo aqui


## Operadores Lógicos

- Producen un resultado booleano (True o False) a partir de expresiones lógicas

- Prioridad:
    1. not
    2. and
    3. or

| Operador | Operación |
| --- | --- |
| and | y (conjunción) |
| or | o (disyunción) |
| not | no (Negación) |

In [ ]:
#El mayor de tres números
a = 10
b = 100
c = 10

if(a>=b and a>=c):
    print("El mayor es a")
    if(a==b or a == c):
        print("Y está repetido")
elif(b>=c and b>=a):
    print("El mayor es b")
    if(b==c or b == a):
        print("Y está repetido")
else:
    print("El mayor es c")
    if(b==c or c == a):
        print("Y está repetido")


In [ ]:
#El mayor de tres números
a = 10
b = 100
c = 10

if(a<=b and a<=c):
    print("El menor es a")
    if(a==b or a == c):
        print("Y está repetido")
elif(b<=c and b<=a):
    print("El menor es b")
    if(b==c or b == a):
        print("Y está repetido")
else:
    print("El menor es c")
    if(b==c or c == a):
        print("Y está repetido")


## **Ejercicio 3:** Solicitar tres números e imprimir el menor


In [ ]:
num1 = float(input("Ingrese un número: "))
num2 = float(input("Ingrese un segundo número: "))
num3 = float(input("Ingrese un tercer número: "))



## Precedencia de operadores

__Recordemos__ que la jerarquía de operadores aritméticos es la siguiente:

1. $()$: parentesis.
2. $**$ : potencia.
3. $* /$ % not : multiplicación, división, módulo, negación (not).
4. $+ -$ and : suma, resta, conjunción (and).
5. $>,<,==,>=,<=,!=,$ or : mayor que, menor que, igual, mayor o igual que, menor o igual que, diferenet, disyunción (or)

## Ejercicios:
1.  Escriba el codigo que a partir de un numero entero devuelva el dia de la semana (0 - 6, de Domingo - Sabado).
2. Diseñar un progrma que reciba 3 variables numericas y devuelva cuantas son positivas, iguales a 0, o negativas.
3. **Detección de fallas en un circuito**:
Escribe un programa que detecte si hay una falla en un circuito basado en la corriente que fluye a través de un resistor. Considere que la corriente normal está entre 0.1 A y 0.5 A. Si la corriente está fuera de este rango, indica que hay una falla.
4. Escribe un programa que controle la temperatura de un sistema. Si la temperatura es mayor a 100°C, enciende un ventilador; si es menor a 50°C, apaga el ventilador; y si está entre 50°C y 100°C, mantén el ventilador en su estado actual.

# Estructura match

Similar a la estructura switch-case de C++.
Ejemplo: Clasificar si un punto esta en el origen o en algun eje.

In [ ]:
coordenada = (0, 1) #Cambiar valores
match coordenada:
    case (0, 0):
        print("Coordenada en origen")
    case (x, 0):
        print(f"Coordenada en eje x: {x}")
    case (0, y):
        print(f"Coordenada en eje y: {y}")
    case (x, y):
        print(f"Coordenada en: {x}, {y}")
    case _:
        print("Error")

Ejercicio: Escribe un programa en Python que simule este control utilizando la instrucción match. El programa debe:
1. Definir una variable `estado_actual`.
2. Recibir entradas de "sensores" (pueden ser variables manuales o input()).
3. Utilizar match `estado_actual` para decidir qué acción tomar y a qué estado cambiar.
Requerimientos Lógicos de Transición:
* Si está en `IDLE` y `boton_inicio` == True $\rightarrow$ Cambiar a `HEATING`.
* Si está en `HEATING` y `temperatura` >= 121 $\rightarrow$ Cambiar a `STERILIZING`.
* Si está en `STERILIZING` y `tiempo_restante` == 0 $\rightarrow$ Cambiar a `COOLING`.
* Si está en `COOLING` y `presion` == 1 $\rightarrow$ Cambiar a `IDLE` e imprimir "Ciclo Completado".